agent/engine/docker_task_engine.go

```diff
@@ -274,48 +278,31 @@ func (engine *DockerTaskEngine) monitorExecAgentProcesses(ctx context.Context) {
 	defer engine.tasksLock.RUnlock()
 	for _, mTask := range engine.managedTasks {
 		task := mTask.Task
-		if !task.IsExecCommandAgentEnabled() {
+		if !task.IsExecCommandAgentEnabled() || task.GetKnownStatus() != apitaskstatus.TaskRunning {
 			continue
 		}
 		for _, c := range task.Containers {
-			go engine.monitorExecAgentRunning(ctx, task.Arn, c)
+			go engine.monitorExecAgentRunning(ctx, task, c)
 		}
 	}
 }
 
 func (engine *DockerTaskEngine) monitorExecAgentRunning(ctx context.Context,
-	tArn string, c *apicontainer.Container) {
-	if c.GetExecCommandAgentMetadata() == nil {
+	task *apitask.Task, c *apicontainer.Container) {
+	if !c.IsRunning() {
 		return
 	}
-	pID := c.GetExecCommandAgentMetadata().PID
-	cName := c.Name
-	cID := c.RuntimeID
-	if cID == "" || pID == "" {
-		return
-	}
-	resp, err := engine.client.TopContainer(ctx, cID, dockerclient.TopContainerTimeout, pID)
+	dockerID, err := engine.getDockerID(task, c)
 	if err != nil {
-		if _, ok := err.(*dockerapi.DockerTimeoutError); ok {
-			seelog.Warnf("Task engine [%s]: Timed out monitoring ExecCommandAgent process for container %s", tArn, cName)
-			return
-		}
-		if strings.Contains(err.Error(), dockerapi.TopProcessNotFoundErrorName) {
-			seelog.Errorf("Task engine [%s]: ExecCommandAgent Process for container %s not running, restarting", tArn, cName)
-			c.SetExecCommandAgentMetadata(&apicontainer.ExecCommandAgentMetadata{})
-			// TODO: [ecs-exec]restart exec agent process for this container
-			return
-		}
-		// TODO: [ecs-exec] consider retry/adding an extra docker exec inspect API call to get more info
-		seelog.Errorf("Task engine [%s]: Error finding ExecCommandAgent process for container %s: %v", tArn, cName, err)
+		seelog.Errorf("Task engine [%s]: Could not retrieve docker id for container", task.Arn)
 		return
 	}
-	// no error, but SSM agent process not found in the container
-	if resp == nil || len(resp.Processes) == 0 {
-		seelog.Errorf("Task engine [%s]: ExecCommandAgent Process not found for container %s, no error, restarting", tArn, cName)
-		c.SetExecCommandAgentMetadata(&apicontainer.ExecCommandAgentMetadata{})
-		// TODO: [ecs-exec]restart exec agent process for this container
+	_, err = engine.execCmdAgentMgr.RestartIfStopped(ctx, engine.client, task, c, dockerID)
+	if err != nil {
+		seelog.Errorf("Task engine [%s]: Failed to restart ExecCommandAgent Process for container [%s]: %v", task.Arn, dockerID, err)
+		// TODO: [ecs-exec] call control plane to report ExecCommandAgent FAIL here
```

Comment:
This is an API that CP is in process of implementing?

Reply:
Yes it is, Also, Ray is gonna work in integrating this part, leaving this TODOs so we have a reference of where to put things

Reply:
(thanks for that)
pretty sure this is going to be an enum, and we'll support `PENDING`, `RUNNING`, `STOPPED` states.